worker/imgproxy.js

```diff
@@ -0,0 +1,194 @@
+  //   - qHD:  960x 540
+  //   - HD:  1280x 720
+  //   - HD+: 1600x 900
+  //   - FHD: 1920x1080
+  //   - QHD: 2560x1440
+  // reference:
+  //   - https://en.wikipedia.org/wiki/Graphics_display_resolution#High-definition_(HD_and_derivatives)
+  //   - https://www.browserstack.com/guide/ideal-screen-sizes-for-responsive-design
+  const resolutions = ['640x360', '960x540', '1280x720', '1600x900', '1920x1080', '2560x1440']
+  const imgproxyUrls = {}
+  for (let url of urls) {
+    if(!url) continue
+
+    console.log("[imgproxy] id:", id, "-- processing url:", url)
+    if (url.startsWith(IMGPROXY_URL)) {
+      console.log("[imgproxy] id:", id, "-- proxy url, decoding original url:", url)
+      // backwards compatibility: we used to replace image urls with imgproxy urls
+      url = decodeOriginalUrl(url)
+      console.log("[imgproxy] id:", id, "-- original url:", url)
+    }
+    if (!(await isImageURL(url, { forceFetch }))) {
+      console.log("[imgproxy] id:", id, "-- not image url:", url)
+      continue
+    }
+    imgproxyUrls[url] = {}
+    for (const res of resolutions) {
+      const [w, h] = res.split('x')
+      const processingOptions = `/rs:fill:${w}:${h}`
+      imgproxyUrls[url][`${w}w`] = createImgproxyUrl(url, processingOptions)
+    }
+  }
+  return imgproxyUrls
+}
+
+const createImgproxyUrl = (url, processingOptions) => {
+  const b64Url = Buffer.from(url, 'utf-8').toString('base64url')
+  const target = `${processingOptions}/${b64Url}`
+  const signature = sign(target)
+  return `${IMGPROXY_URL}${signature}${target}`
+}
+
+async function fetchWithTimeout (resource, { timeout = 1000, ...options } = {}) {
+  const controller = new AbortController()
+  const id = setTimeout(() => controller.abort(), timeout)
+
+  const response = await fetch(resource, {
+    ...options,
+    signal: controller.signal
+  })
+  clearTimeout(id)
+
+  return response
+}
+
+const isImageURL = async (url, { forceFetch }) => {
+  if (cache.has(url)) return cache.get(url)
+
+  if (!forceFetch && matchUrl(imageUrlMatchers, url)) {
+    return true
+  }
+  if (!forceFetch && matchUrl(exclude, url)) {
+    return false
+  }
+
+  let isImage
+
+  // first run HEAD with small timeout
+  try {
+    // https://stackoverflow.com/a/68118683
+    const res = await fetchWithTimeout(url, { timeout: 1000, method: 'HEAD' })
+    const buf = await res.blob()
+    isImage = buf.type.startsWith('image/')
+  } catch (err) {
+    console.log(url, err)
+  }
+
+  // For HEAD requests, positives are most likely true positives.
+  // However, negatives may be false negatives
+  if (isImage) {
+    cache.set(url, true)
+    return true
+  }
+
+  // if not known yet, run GET request with longer timeout
+  try {
+    const res = await fetchWithTimeout(url, { timeout: 10000 })
+    const buf = await res.blob()
+    isImage = buf.type.startsWith('image/')
+  } catch (err) {
+    console.log(url, err)
+  }
+
+  cache.set(url, isImage)
+  return isImage
+}
```

Comment:
How frequently is this needed? Are there images that take 10s to return a HEAD request?

Reply:
Some hosts don't support HEAD at all or respond with wrong headers.

For example, zapread.com (which is in the top 5 hosts used by stackers) responds with 404:

```
curl -I https://www.zapread.com/i/9QP
HTTP/2 404
date: Sat, 30 Sep 2023 20:45:06 GMT
content-type: text/html; charset=utf-8
access-control-expose-headers: Request-Context
cache-control: private
set-cookie: ARRAffinity=24410e45ca5e57f20082acb1c6e848971a37deecc97f93db2394281f8d4f09c7;Path=/;HttpOnly;Secure;Domain=www.zapread.com
set-cookie: ARRAffinitySameSite=24410e45ca5e57f20082acb1c6e848971a37deecc97f93db2394281f8d4f09c7;Path=/;HttpOnly;SameSite=None;Secure;Domain=www.zapread.com
strict-transport-security: max-age=31536000; includeSubDomains
x-aspnet-version: 4.0.30319
request-context: appId=cid-v1:1f8ce37e-229f-450c-a0fd-1303f2fc9b8c
x-powered-by: ASP.NET
x-content-type-options: nosniff
referrer-policy: no-referrer
cf-cache-status: DYNAMIC
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=sARzDZKmKzHoC8OHea0UssbRwy4EL%2FFI0Fl8vZB21iqoPgFawu7lSPePyri24k9AQeUX6lvGjZ1UDBob%2BVtjEj6FUj5Pbvv4G0Xw20i0xPwpLBm0tcap%2Br9j0x1PCm7%2BL1w%3D"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
server: cloudflare
cf-ray: 80ef4ec20918421d-EWR
alt-svc: h3=":443"; ma=86400
```

But if you go to https://www.zapread.com/i/9QP, there is an image.

So for reliable image detection, a GET request is definitely needed imo.

Also, you're looking at the GET request here. The timeout for HEAD requests is 1s.